### PR TITLE
batch saving: split up persist_notification function

### PR DESCRIPTION
**Pull request by sastels**
# Summary | Résumé

https://app.zenhub.com/workspaces/notify-planning-614b3ad91bc2030015ed22f5/issues/cds-snc/notification-planning/149

Split `persist_notification()` into two functions (note: keeping the original there as well):
- `transform_notification()` that takes the same params as `persist_notification()` and returns a notification
- `db_save_notification()` which saves a notification to the database

Change the `/v2/notifications/email` and `/v2/notifications/sms` endpoints to use the two new functions instead of `persist_notification()`.

This will allow us to next replace `db_save_notification()` with something that pushes the notification to redis.

# Test instructions | Instructions pour tester la modification

tested locally by posting to the `/v2/notifications/email` and `/v2/notifications/sms` endpoints

# Release Instructions | Instructions pour le déploiement

None. (in particular, we are still doing the same steps, just in two functions, so there isn't a feature flag yet)

# Reviewer checklist | Liste de vérification du réviseur

This is a suggested checklist of questions reviewers might ask during their
review | Voici une suggestion de liste de vérification comprenant des questions
que les réviseurs pourraient poser pendant leur examen :

- [ ] Is the code maintainable? | Est-ce que le code peut être maintenu?
- [ ] Have you tested it? | L’avez-vous testé?
- [ ] Are there automated tests? | Y a-t-il des tests automatisés?
- [ ] Does this cause automated test coverage to drop? | Est-ce que ça entraîne
      une baisse de la quantité de code couvert par les tests automatisés?
- [ ] Does this break existing functionality? | Est-ce que ça brise une
      fonctionnalité existante?
- [ ] Does this change the privacy policy? | Est-ce que ça entraîne une
      modification de la politique de confidentialité?
- [ ] Does this introduce any security concerns? | Est-ce que ça introduit des
      préoccupations liées à la sécurité?
- [ ] Does this significantly alter performance? | Est-ce que ça modifie de
      façon importante la performance?
- [ ] What is the risk level of using added dependencies? | Quel est le degré de
      risque d’utiliser des dépendances ajoutées?
- [ ] Should any documentation be updated as a result of this? (i.e. README
      setup, etc.) | Faudra-t-il mettre à jour la documentation à la suite de ce
      changement (fichier README, etc.)?

### CODE REVIEW

File: tests/app/v2/notifications/test_post_letter_notifications.py [hunk 1]
```diff
@@ -278,7 +278,7 @@ def test_notification_returns_400_if_address_doesnt_have_underscores(client, sam
 
 
 def test_returns_a_429_limit_exceeded_if_rate_limit_exceeded(client, sample_letter_template, mocker):
-    persist_mock = mocker.patch("app.v2.notifications.post_notifications.persist_notification")
+    persist_mock = mocker.patch("app.notifications.process_notifications.persist_notification")
```
Comment (sastels) [inline tests/app/v2/notifications/test_post_letter_notifications.py:281]: Oddly they were mocking another module that imported `persist_notification` rather that the module where `persist_notification` was defined? I guess it all goes back the same function? 🤷

**Review by jzbahrai — APPROVED** [review]

File: tests/app/notifications/test_process_notification.py [hunk 2]
```diff
@@ -654,3 +656,299 @@ def test_persist_notifications_list(self, sample_job, sample_api_key, notify_db_
...
+            personalisation={},
+            notification_type="sms",
+            api_key_id=sample_api_key.id,
+            key_type=sample_api_key.key_type,
+            reference="ref2",
+        )
+
+        mock_incr.assert_called_once_with(
+            str(sample_template.service_id) + "-2016-01-01-count",
+        )
```
Comment (jimleroyer) [inline tests/app/notifications/test_process_notification.py:954]: Very nice set of tests and more targeted between transform and db saving.

File: tests/app/v2/notifications/test_post_letter_notifications.py (hunk 1 above)
Comment (jimleroyer) [reply tests/app/v2/notifications/test_post_letter_notifications.py:281]: 🤯

**Comment by josh-cds** [comment]
@jimleroyer has this been QA'd?

**Comment by jimleroyer** [comment]
@josh-cds This is another pull request and not an issue per se. Let's discuss offline what to do with these. 😄

**Pull request merged**